File: rad/engine/02_catalog/store/transition_cleanup.go
package store

import (
	"context"

	kv "github.com/Southclaws/rad/rad/engine/01_kv"
	"github.com/Southclaws/rad/rad/engine/02_catalog/model"
)

// TransitionCompactionEligible separates terminal diagnostic retention from
// physical cleanup. A detailed terminal record can be compacted only after its
// typed cleanup checkpoint is durable and no diagnostic pin retains it.
func TransitionCompactionEligible(
	ctx context.Context,
	view kv.KV,
	transition model.SchemaTransition,
) (bool, error) {
	if !transition.CompactedAt.IsZero() {
		return false, nil
	}
	if transition.Kind == model.TransitionIndexBuild &&
		transition.IndexRequest != nil &&
		len(transition.Index.ColumnIDs) == 0 {
		switch transition.State {
		case model.TransitionReady, model.TransitionCancelled, model.TransitionFailed:
			pinned, err := TransitionDiagnosticsPinned(ctx, view, transition.ID)
			return !pinned, err
		}
	}
	if transition.Kind == model.TransitionColumnReplacement &&
		transition.ColumnReplacement == nil {
		switch transition.State {
		case model.TransitionReady, model.TransitionCancelled, model.TransitionFailed:
			pinned, err := TransitionDiagnosticsPinned(ctx, view, transition.ID)
			return !pinned, err
		}
	}
	var reclamationID string
	switch transition.State {
	case model.TransitionReady:
		switch transition.Kind {
		case model.TransitionIndexBuild:
			reclamationID = TransitionDeltaReclamationID(transition.ID)
		case model.TransitionColumnReplacement:
			reclamationID = ReplacedColumnReclamationID(transition.ID)
		case model.TransitionConstraintValidation:
			reclamationID = ConstraintValidationReclamationID(transition.ID)
		}
	case model.TransitionCancelled:
		switch transition.Kind {
		case model.TransitionIndexBuild:
			reclamationID = CancelledIndexReclamationID(transition.ID)
		case model.TransitionColumnReplacement:
			reclamationID = CancelledReplacementReclamationID(transition.ID)
		case model.TransitionConstraintValidation:
			reclamationID = ConstraintValidationReclamationID(transition.ID)
		}
	case model.TransitionFailed:
		switch transition.Kind {
		case model.TransitionIndexBuild:
			reclamationID = FailedIndexReclamationID(transition.ID)
		case model.TransitionColumnReplacement:
			reclamationID = FailedReplacementReclamationID(transition.ID)
		case model.TransitionConstraintValidation:
			reclamationID = ConstraintValidationReclamationID(transition.ID)
		}
	default:
		return false, nil
	}
	if reclamationID == "" {
		return false, nil
	}
	pinned, err := TransitionDiagnosticsPinned(ctx, view, transition.ID)
	if err != nil || pinned {
		return false, err
	}
	reclamation, ok, err := GetReclamation(ctx, view, reclamationID)
	if err != nil || !ok {
		return false, err
	}
	return reclamation.State == model.ReclamationReclaimed, nil
}

File: src/engine/02_catalog/store/transition_cleanup.rs
use crate::engine::catalog::Result;
use crate::engine::catalog::model::{
    ReclamationState, SchemaTransition, TransitionKind, TransitionState,
};
use crate::engine::kv::KvView;

use super::{
    cancelled_index_reclamation_id, cancelled_replacement_reclamation_id,
    constraint_validation_reclamation_id, failed_index_reclamation_id,
    failed_replacement_reclamation_id, get_reclamation, replaced_column_reclamation_id,
    transition_delta_reclamation_id, transition_diagnostics_pinned,
};

pub async fn transition_compaction_eligible<V: KvView + ?Sized>(
    view: &mut V,
    transition: &SchemaTransition,
) -> Result<bool> {
    if !transition.compacted_at.is_zero() {
        return Ok(false);
    }
    let terminal = transition.state.is_terminal();
    let legacy_without_cleanup = (transition.kind == TransitionKind::IndexBuild
        && transition.index_request.is_some()
        && transition.index.column_ids.is_empty())
        || (transition.kind == TransitionKind::ColumnReplacement
            && transition.column_replacement.is_none());
    if terminal && legacy_without_cleanup {
        return Ok(!transition_diagnostics_pinned(view, &transition.id).await?);
    }

    let reclamation_id = match (transition.state, transition.kind) {
        (TransitionState::Ready, TransitionKind::IndexBuild) => {
            transition_delta_reclamation_id(&transition.id)
        }
        (TransitionState::Ready, TransitionKind::ColumnReplacement) => {
            replaced_column_reclamation_id(&transition.id)
        }
        (TransitionState::Ready, TransitionKind::ConstraintValidation)
        | (TransitionState::Cancelled, TransitionKind::ConstraintValidation)
        | (TransitionState::Failed, TransitionKind::ConstraintValidation) => {
            constraint_validation_reclamation_id(&transition.id)
        }
        (TransitionState::Cancelled, TransitionKind::IndexBuild) => {
            cancelled_index_reclamation_id(&transition.id)
        }
        (TransitionState::Cancelled, TransitionKind::ColumnReplacement) => {
            cancelled_replacement_reclamation_id(&transition.id)
        }
        (TransitionState::Failed, TransitionKind::IndexBuild) => {
            failed_index_reclamation_id(&transition.id)
        }
        (TransitionState::Failed, TransitionKind::ColumnReplacement) => {
            failed_replacement_reclamation_id(&transition.id)
        }
        _ => return Ok(false),
    };
    if transition_diagnostics_pinned(view, &transition.id).await? {
        return Ok(false);
    }
    Ok(get_reclamation(view, &reclamation_id)
        .await?
        .is_some_and(|reclamation| reclamation.state == ReclamationState::Reclaimed))
}
